gamla/transducer.py

```diff
@@ -0,0 +1,68 @@
+import functools
+from typing import Any, Callable, Iterable, TypeVar
+
+import toolz
+from toolz import curried
+
+from gamla import functional, functional_generic
+
+_S = TypeVar("_S")
+_X = TypeVar("_X")
+Reducer = Callable[[_S, _X], _S]
+Transducer = Callable[[Reducer], Reducer]
+
+
+def transduce(transformation, step, initial, collection):
+    return functools.reduce(transformation(step), collection, initial)
+
+
+def _transform_by_key(injector):
+    return lambda key, reducer: lambda step: lambda s, x: step(
+        injector(s, key, reducer(s[key], x)),
+        x,
+    )
+
+
+def _replace_index(the_tuple, index, value):
+    return (*the_tuple[:index], value, *the_tuple[index + 1 :])
+
+
+apply_spec = functional_generic.compose_left(
+    dict.items,
+    curried.map(functional.star(_transform_by_key(toolz.assoc))),
```

Comment:
gamla.map ?

Reply:
i'm not sure we need the async pyrotechnics here just yet, so want to keep the dependencies in other parts of the lib minimal.